runtime/src/registrar.rs

```diff
@@ -0,0 +1,1346 @@
+
+			if PendingSwap::get(other) == Some(id) {
+				// actually do the swap.
+				T::OnSwap::ensure_can_swap(id, other)?;
+
+				Self::force_unschedule(|i| i == id || i == other);
+				Parachains::mutate(|ids| swap_ordered_existence(ids, id, other));
+				Paras::mutate(id, |i|
+					Paras::mutate(other, |j|
+						rstd::mem::swap(i, j)
+					)
+				);
+
+				let _ = T::OnSwap::on_swap(id, other);
+			} else {
+				PendingSwap::insert(id, other);
+			}
+		}
+
+		/// Block initializer. Clears SelectedThreads and constructs/replaces Active.
+		fn on_initialize() {
+			let next_up = SelectedThreads::mutate(|t| {
+				let r = if t.len() >= QUEUE_SIZE {
+					// Take the first set of parachains in queue
+					t.remove(0)
+				} else {
+					vec![]
+				};
+				if t.len() < QUEUE_SIZE {
+					t.push(vec![]);
+				}
+				r
+			});
+			// mutable so that we can replace with `None` if parathread appears in new schedule.
+			let mut retrying = Self::take_next_retry();
+			if let Some(((para, _), _)) = retrying {
+				// this isn't really ideal: better would be if there were an earlier pass that set
+				// retrying to the first item in the Missed queue that isn't already scheduled, but
+				// this is potentially O(m*n) in terms of missed queue size and parathread pool size.
+				if next_up.iter().any(|x| x.0 == para) {
+					retrying = None
+				}
+			}
+
+			let paras = Parachains::get().into_iter()
+				.map(|id| (id, None))
+				.chain(next_up.into_iter()
+					.map(|(para, collator)| (para, Some((collator, Retriable::WithRetries(0)))))
+				).chain(retrying.into_iter()
+					.map(|((para, collator), retries)| (para, Some((collator, Retriable::WithRetries(retries + 1)))))
+				).collect::<Vec<_>>();
+			Active::put(paras);
+		}
+
+		fn on_finalize() {
+			// a block without this will panic, but let's not panic here.
+			if let Some(proceeded_vec) = parachains::DidUpdate::get() {
+				// We can't remove the non-parathread items as with Active, so we reverse everything
+				// so that we safely ignore permanent chains (which are at the beginning of
+				// `DidUpdate`).
+				let mut proceeded = proceeded_vec.into_iter().rev();
+				let mut i = proceeded.next();
+
+				let active = Active::get().into_iter()
+					.rev()
+					// Early exit to skip any permanent parachains (i.e. without collators). Not
+					// needed with the following line, but early-exit should provide some speed-up.
+					.take_while(|i| i.1.is_some())
+					.filter_map(|i| match i.1 {
+						None | Some((_, Retriable::Never)) => None,
+						Some((c, Retriable::WithRetries(n))) => Some((i.0, c, n)),
+					});
+
+				// Check all normally scheduled parathreads: the rest of the `active` iterator are
+				// just normally scheduled parathreads (albeit in reverse order).
+				for sched in active {
+					match i {
+						// Scheduled parachain proceeded properly. Move onto next item.
+						Some(para) if para == sched.0 => i = proceeded.next(),
```

Comment:
This makes a few assumptions:
  1. that `active` is in ascending (well, from this perspective, descending, since we've reversed it) order by para ID. it doesn't seem guaranteed by `on_initialize`.
  2. that `proceeded` has all parathreads at the end (not guaranteed by interface or implementation fo `parachains::set_heads`)

Reply:
indeed parathread IDs are not guaranteed to be greater than parachain IDs (you're right about the swap mechanism precluding this).

however `Active` is (now) documented as "Ordered as Parachains (ordered by ascending ID) ++ Selected_Parathreads (the last entry is always the item from the retry queue, if one exists)". as such only the first part is sorted. the trick with reversing it is so that we can quickly (O(N)) iterate through the parathreads and figure out whether they were progressed or not.

Reply:
Resolved in my latest commit.